lib/spack/spack/test/cmd/env.py

```diff
@@ -3119,6 +3120,46 @@ def test_environment_depfile_makefile(depfile_flags, expected_installs, tmpdir,
     assert len(specs_that_make_would_install) == len(expected_installs)
 
 
+def test_environment_depfile_spec_format_special_chars(tmpdir, mock_packages, monkeypatch):
+    env("create", "test")
+    makefile = str(tmpdir.join("Makefile"))
+    with ev.read("test"):
+        add("dttop")
+        concretize()
+
+    orig_format = spack.spec.Spec.format
+
+    def _contrived_spec_format(obj, format_str):
+        if obj.name == "dtlink1":
+            return "dtlink1-githash=version-spackhash"
+        else:
+            return orig_format(obj, format_str)
+
+    monkeypatch.setattr(spack.spec.Spec, "format", _contrived_spec_format)
```

Comment:
Since you already have mock_packages, can't you actually use a normal git version with user specified spack version to the same effect? That would be less contrived

Reply:
I wanted to avoid patching the packages to achieve this condition: IMO the source of all errors like the one this PR aims to correct is specially-interpreted characters via `Spec.format`; therefore I argue that this is the least-contrived way to test the outcome (changing things about the versions etc. would silently render this test useless if `spec.format` happened to behave differently in some scenarios vs. others).

Reply:
Hm, it's a rule of thumb not to mock in integration tests, cause assumptions on the implementation make the test redundant on refactor. If you want an integration test for git versions, then just use a git version. If you want to test normalization of targets, then write a unit test. In both cases you don't need to use monkeypatch

Reply:
Ah... I see what the issue is, or at least I think I do :p we're generating target/prereqs without rules like

```
<pkg a>-<version a>-<hash a>: <pkg b>-<version b>-<hash b>
```

so presumably if version contains an `=` it parses as a variable declaration

Reply:
>  If you want to test normalization of targets

I don't want to normalize the entire target: the depfile uses special characters in targets for their intended effect. The issue is when those targets are partially defined using `Spec.format`, which introduces *unexpected* special characters.

> If you want an integration test for git versions, then just use a git version.

I don't think git versions are the (real) problem. For a system that generates makefiles of spack installs and needs to create targets based on the Specs, each such target is composed of:

* Static components relevant to makefile processing
* Details extracted from the Spec itself

The latter is the only place that can surprise us. The current logic has no formalized abstraction for the latter and directly uses `.format`. I considered it appropriate to just alter `Spec.format`, but it would be more-robust to mock the Spec object entirely.

> Hm, it's a rule of thumb not to mock in integration tests, cause assumptions on the implementation make the test redundant on refactor.

given the above, while I'd say it is strictly true that this mocking makes an assumption in the implementation, setting a git version is even more a violation of this rule (conditional changes to `str()` of version could render this test useless: why not target the point where the depfile logic directly interfaces with the Spec API?).

Reply:
I'm sorry, but I'd rather see `monkeypatch` dropped. The other point is that `Spec.format` is so commonly called, mocking it will likely break invariants elsewhere in the future.

Please just add an integration/regression test with depfile and git versions which is the problem you're solving in the PR, and if you don't like integration tests, refactor so that you can unit test the adjacency list